examples/pytorch/dimenet/main.py

```diff
@@ -0,0 +1,254 @@
+    lengths = [num_train, num_valid, num_data - num_train - num_valid]
+    if shuffle:
+        indices = np.random.RandomState(seed=random_state).permutation(num_data)
+    else:
+        indices = np.arange(num_data)
+    return [Subset(dataset, indices[offset - length:offset]) for offset, length in zip(accumulate(lengths), lengths)]
+
+@torch.no_grad()
+def ema(ema_model, model, decay):
+    msd = model.state_dict()
+    for k, ema_v in ema_model.state_dict().items():
+        model_v = msd[k].detach()
+        ema_v.copy_(ema_v * decay + (1. - decay) * model_v)
+
+def edge_init(edges):
+    R_src, R_dst = edges.src['R'], edges.dst['R']
+    dist = torch.sqrt(F.relu(torch.sum((R_src - R_dst) ** 2, -1)))
+    # d: bond length, o: bond orientation
+    return {'d': dist, 'o': R_src - R_dst}
+
+def _collate_fn(batch):
+    graphs, line_graphs, labels = map(list, zip(*batch))
+    g, l_g = dgl.batch(graphs), dgl.batch(line_graphs)
+    labels = torch.tensor(labels, dtype=torch.float32)
+    return g, l_g, labels
+
+def train(device, model, opt, loss_fn, train_loader):
+    model.train()
+    epoch_loss = 0
+    num_samples = 0
+
+    for g, l_g, labels in train_loader:
+        g = g.to(device)
+        l_g = l_g.to(device)
+        labels = labels.to(device)
+        logits = model(g, l_g)
+        loss = loss_fn(logits, labels.view([-1, 1]))
+        epoch_loss += loss.data.item() * len(labels)
+        num_samples += len(labels)
+        opt.zero_grad()
+        loss.backward()
+        opt.step()
+
+    return epoch_loss / num_samples
+
+@torch.no_grad()
+def evaluate(device, model, valid_loader):
+    model.eval()
+    predictions_all, labels_all = [], []
+    
+    for g, l_g, labels in valid_loader:
+        g = g.to(device)
+        l_g = l_g.to(device)
+        logits = model(g, l_g)
+        labels_all.extend(labels)
+        predictions_all.extend(logits.view(-1,).cpu().numpy())
+    
+    return np.array(predictions_all), np.array(labels_all)
+
+@click.command()
+@click.option('-m', '--model-cnf', type=click.Path(exists=True), help='Path of model config yaml.')
+def main(model_cnf):
+    yaml = YAML(typ='safe')
+    model_cnf = yaml.load(Path(model_cnf))
+    model_name, model_params, train_params, pretrain_params = model_cnf['name'], model_cnf['model'], model_cnf['train'], model_cnf['pretrain']
+    logger.info(f'Model name: {model_name}')
+    logger.info(f'Model params: {model_params}')
+    logger.info(f'Train params: {train_params}')
+
+    if model_params['targets'] in ['mu', 'homo', 'lumo', 'gap', 'zpve']:
+        model_params['output_init'] = nn.init.zeros_
+    else:
+        # 'GlorotOrthogonal' for alpha, R2, U0, U, H, G, and Cv
+        model_params['output_init'] = GlorotOrthogonal
+
+    logger.info('Loading Data Set')
+    dataset = QM9(label_keys=model_params['targets'], edge_funcs=[edge_init])
+
+    # data split
+    train_data, valid_data, test_data = split_dataset(dataset,
```

Comment:
Since this is a random split, are we using the same split as the one used by authors?

Reply:
yep